drv/fpga-server/src/main.rs

```diff
@@ -0,0 +1,380 @@
+                program_n: sys_api::Port::B.pin(10),
+                configuration_port,
+                user_design,
+                user_design_reset_n: sys_api::Port::D.pin(11),
+                user_design_reset_duration: ecp5::USER_DESIGN_RESET_DURATION,
+            };
+        } else {
+            compile_error!("Board is not supported by the task/fpga");
+        }
+    }
+    driver.configure_gpio();
+
+    let device = Ecp5::new(driver);
+    let mut incoming = [0u8; idl::INCOMING_SIZE];
+    let mut server = ServerImpl {
+        lock_holder: None,
+        device: &device,
+        buffer: [0u8; 128],
+        bitstream_loader: None,
+    };
+
+    if let Ok(DeviceState::AwaitingBitstream) = server.device.device_state() {
+        ringbuf_entry!(Trace::DeviceId(server.device.device_id().unwrap()));
+    }
+
+    loop {
+        idol_runtime::dispatch(&mut incoming, &mut server);
+    }
+}
+
+enum BitstreamLoader<'a, Device: Fpga<'a>> {
+    Uncompressed(Device::Bitstream, usize),
+    Compressed(gnarle::Decompressor, Device::Bitstream, usize),
+}
+
+struct ServerImpl<'a, Device: Fpga<'a> + FpgaUserDesign> {
+    lock_holder: Option<userlib::TaskId>,
+    device: &'a Device,
+    buffer: [u8; 128],
+    bitstream_loader: Option<BitstreamLoader<'a, Device>>,
+}
+
+/// This UserDesignLock is used to ensure atomic read/write operations to the
+/// user design port.
+///
+/// TODO (arjen): This should probably move into the FpgaUserDesign trait, but
+/// the semantics are a bit messy. This will do the correct thing for now, until
+/// there is a bit more time to consider the semantics of the SPI driver and
+/// what is intended to happen when an FPGA API user wants to lock the user
+/// design.
+struct UserDesignLock<'a, Device: FpgaUserDesign>(&'a Device);
+
+impl<Device: FpgaUserDesign> Drop for UserDesignLock<'_, Device> {
+    fn drop(&mut self) {
+        self.0.user_design_release().unwrap()
+    }
+}
+
+impl<'a, Device: Fpga<'a> + FpgaUserDesign> ServerImpl<'a, Device> {
+    fn lock_user_design(
+        &self,
+    ) -> Result<UserDesignLock<'a, Device>, FpgaError> {
+        self.device.user_design_lock().map_err(FpgaError::from)?;
+        Ok(UserDesignLock(self.device))
+    }
+}
+
+type RequestError = idol_runtime::RequestError<FpgaError>;
+type ReadDataLease = LenLimit<Leased<R, [u8]>, 128>;
+type WriteDataLease = LenLimit<Leased<W, [u8]>, 128>;
+
+impl<'a, Device: Fpga<'a> + FpgaUserDesign> idl::InOrderFpgaImpl
+    for ServerImpl<'a, Device>
+{
+    fn recv_source(&self) -> Option<userlib::TaskId> {
+        self.lock_holder
+    }
+
+    fn closed_recv_fail(&mut self) {
+        // Welp, someone had asked us to lock and then died. Release the
+        // lock and any resources acquired from the device driver.
+        self.lock_holder = None;
+        self.bitstream_loader = None;
+    }
+
+    fn lock(&mut self, msg: &userlib::RecvMessage) -> Result<(), RequestError> {
+        if let Some(task) = self.lock_holder {
```

Comment:
Do we actually expect tasks to lock the server multiple times? It seems more intuitive to return an error if the `lock()` function is called when the server is already locked.

Reply:
I added an error for this.